### PR TITLE
feat(portal): Use Service Accounts to sync Google Workspace directory

### PR DESCRIPTION
**Pull request by AndrewDryga**
We will need to update the docs for the website, some screens to show where the settings are:
Screenshot 2024-08-19 at 1 04 23 PM
Screenshot 2024-08-19 at 1 04 02 PM

Related [#5959](https://github.com/firezone/firezone/issues/5959)

### CODE REVIEW

File: elixir/apps/domain/lib/domain/auth/adapters/google_workspace/api_client.ex [hunk 1]
```diff
@@ -34,6 +34,50 @@ defmodule Domain.Auth.Adapters.GoogleWorkspace.APIClient do
     [conn_opts: [transport_opts: transport_opts]]
   end
 
+  # curl -d 'grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Ajwt-bearer&assertion=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJodHRwczovL29hdXRoMi5nb29nbGVhcGlzLmNvbS90b2tlbiIsImV4cCI6MTcyNDE5NTQ5MCwiaWF0IjoxNzI0MTkxODkwLCJpc3MiOiJmaXJlem9uZS1pZHAtc3luY0BvcGVuaWQtY29ubmVjdC10ZXN0LTM5MTcxOS5pYW0uZ3NlcnZpY2VhY2NvdW50LmNvbSIsInNjb3BlIjoib3BlbmlkIGVtYWlsIHByb2ZpbGUgaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vYXV0aC9hZG1pbi5kaXJlY3RvcnkuY3VzdG9tZXIucmVhZG9ubHkgaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vYXV0aC9hZG1pbi5kaXJlY3Rvcnkub3JndW5pdC5yZWFkb25seSBodHRwczovL3d3dy5nb29nbGVhcGlzLmNvbS9hdXRoL2FkbWluLmRpcmVjdG9yeS5ncm91cC5yZWFkb25seSBodHRwczovL3d3dy5nb29nbGVhcGlzLmNvbS9hdXRoL2FkbWluLmRpcmVjdG9yeS51c2VyLnJlYWRvbmx5In0.hFZl772n74blSgJ2dsZiLKr72tZsDFazuLQ4dVEWBGeR-EEtrhkCA5D7FFuKSWQiW9fbu63LQlO-YiqICSuZv0BGkVzzfZ2Gs96PbGf5ezagq3SJA8H_pbo95Mmd2J_cLwn97NmlfXaksdSQgGdhGi_NtUFRavR_A7idvZWQIQyyt96l5eZULU1QeEm61zx6QFHvx8HiuYYqWurBbeg9_quiSsHJ2yXWRCAg98bEBu0swCV68uBzjkaQuAE-zCqnoDQdfVZCf4-11qoTs-_mgUURW86JhGAWjSKbrDCtHteprUgYRuhpxRsOt4TeK0cbIQFTP1eHxNOmOGfupM34fQ' https://oauth2.googleapis.com/token
```
Comment (bmanifold) [inline elixir/apps/domain/lib/domain/auth/adapters/google_workspace/api_client.ex]: Guessing this was a debugging comment accidentally left in?

Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth/adapters/google_workspace/api_client.ex]: Good catch, thanks

**Review by jamilbk — COMMENTED** [review]
Will defer to @bmanifold on this one. I don't think there's too many user-facing changes other than the docs changes needed. 

Tracking that in the launch here:

https://github.com/firezone/firezone/issues/6048

**Comment by jamilbk** [comment]
Merging `main` to get checks to pass

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 2]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
             </p>
           </:content>
         </.step>
 
         <.step>
-          <:title>Step 6. Configure Firezone</:title>
+          <:title>Step 6: Create service account with domain-wide delegation</:title>
+          <:content>
+            <p class="mb-4">
+              Go the
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              Go to the
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 3]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+                href="https://console.cloud.google.com/iam-admin/serviceaccounts"
+                target="_blank"
+                class={link_style()}
+              >
+                <strong>Service Accounts</strong>
+              </a>
+              page of the Google Cloud Console and click <strong>Create Service Account</strong>.
+            </p>
+            <p class="mb-4">
+              Click on the created service account, then click <strong>Keys</strong>
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              Click on the created service account, then click the <strong>Keys</strong>
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 4]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+              tab, <strong>Add Key</strong>
+              and select <strong>Create new key</strong>. Select <strong>JSON</strong>
+              and click <strong>Create</strong>. The contents of the downloaded JSON will be used for
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              and click <strong>Create</strong>. The contents of the downloaded JSON will be used for the
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 5]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+              <strong>Service Account JSON Key</strong>
+              field of the form below.
+            </p>
+            <p class="mb-4">
+              Go back to <strong>Details</strong>
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              Go back to the <strong>Details</strong>
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 6]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+              tab and copy <strong>Unique ID</strong>
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              tab and copy the <strong>Unique ID</strong>
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 7]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+              (OAuth 2 Client ID), you will need it for the next step.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              (OAuth 2 Client ID). You will need it for the next step.
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 8]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+            </p>
+            <p class="mb-4">
+              Sign into the Google Workspace Admin Console and go to <strong>Security</strong>, then <strong>Access and Data Control</strong>, and then <a
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              Sign in to the Google Workspace Admin Console and go to <strong>Security</strong>, then <strong>Access and Data Control</strong>, and then <a
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 9]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+                href="https://admin.google.com/ac/owl"
+                target="_blank"
+                class={link_style()}
+              ><strong>API Controls</strong></a>. Click <strong>Manage Domain Wide Delegation</strong>,
+              <strong>Add new</strong>
+              and paste <strong>Unique ID</strong>
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              and paste the <strong>Unique ID</strong>
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 10]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+              from previous step to <strong>Client ID</strong>
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              from the previous step to the <strong>Client ID</strong>
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 11]
```diff
@@ -152,13 +166,61 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+              field and add the following scopes: <.code_block
+                id="oauth-scopes"
+                class="w-full text-xs mb-4 whitespace-pre-line rounded"
+                phx-no-format
+              ><%= Enum.join(GoogleWorkspace.Settings.scope(), "\n") %></.code_block>
+            </p>
+            <p class="mb-4">
+              Then click <strong>Authorize</strong>.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
              Finally, click <strong>Authorize</strong>.
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 12]
```diff
@@ -185,7 +247,7 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
                     required
                   />
                   <p class="mt-2 text-xs text-neutral-500">
-                    The Client ID from the previous step.
+                    The Client ID from the Step 5.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
                    The Client ID from Step 5.
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 13]
```diff
@@ -197,7 +259,41 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
                     required
                   />
                   <p class="mt-2 text-xs text-neutral-500">
-                    The Client secret from the previous step.
+                    The Client secret from the Step 5.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
                    The Client secret from Step 5.
```

File: elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex [hunk 14]
```diff
@@ -197,7 +259,41 @@ defmodule Web.Settings.IdentityProviders.GoogleWorkspace.Components do
...
+                          |> Jason.encode!()
+
+                        binary when is_binary(binary) ->
+                          binary
+                      end
+                    }
+                    required
+                  />
+                  <p class="mt-2 text-xs text-neutral-500">
+                    The Service Account JSON Key from the Step 6.
```
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/settings/identity_providers/google_workspace/components.ex]:
```suggestion
                    The Service Account JSON Key from Step 6.
```

**Review by jamilbk — APPROVED** [review]
Just a bunch of small grammar fixes otherwise LGTM 🚀 

Wondering if we should prioritize doing this for the other IdPs?

**Comment by AndrewDryga** [comment]
@bmanifold trying to get it tested in staging, we can address PR review separately

**Comment by bmanifold** [comment]
> @bmanifold trying to get it tested in staging, we can address PR review separately

👍 Sounds good

**Pull request merged**